hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/HBaseInterClusterReplicationEndpoint.java

```diff
@@ -133,12 +152,12 @@ public void init(Context context) throws IOException {
     // TODO: This connection is replication specific or we should make it particular to
     // replication and make replication specific settings such as compression or codec to use
     // passing Cells.
-    this.conn = (ClusterConnection) ConnectionFactory.createConnection(this.conf);
+    this.conn = (ClusterConnection) createConnection(this.conf);
```

Comment:
Do we need force createConnection() to return a connection of type ClusterConnection? Otherwise, theoretically overrides can return any type of Connection (may or may not be ClusterConnection) and the type cast here fails. Just a thought.

Reply:
Yeah, that kind of restrict the extensibility, which is what we are trying to improve here, but addressing that would require an extra refactoring effort, just as @joshelser mentioned on the master PR. For example, in master branch, we are forced to return AsyncClusterConnection.

Reply:
> that kind of restrict the extensibility, which is what we are trying to improve here,

Hmm.. May be I'm missing something, what does your use case look like here where you have an override that does a generic Connection rather than (some type of) ClusterConnection? I'm curious how it restricts extensibility because to use this, it has to be a ClusterConnection, otherwise it fails at runtime, no?

Reply:
> what does your use case look like here where you have an override that does a generic Connection rather than (some type of) ClusterConnection?

I have a _HBaseInterClusterReplicationEndpoint_ extension where I'm still using _ConnectionFactory_ for creating connections, but I use different factory method than what is used by _HBaseInterClusterReplicationEndpoint_. I don't really care about which connection implementation the factory is returning, I just rely on whatever is the default (in 2.x, it's _ClusterConnection_). I noticed, though, that default connection type changed between 2.x and 3.0, from _ClusterConnection_ to _AsycnClusterConnection_. So, if we keep _create_ methods referencing generic _Connection_ only, I won't need to change my _HBaseInterClusterReplicationEndpoint_ extension if I upgrade to hbase 3.x (provided, of course, that _HBaseInterClusterReplicationEndpoint.conn_ variable also changes to same default returned by  _ConnectionFactory_).

Reply:
Oh I see.. I _think_ I get the problem. Basically you are just relying on ConnectionFactory and not worried about the type it is returning. Ya, refactoring it will be a bigger patch.

nit: How about adding a preconditions check.. Preconditions.checkState(returnedConn instanceof ClusterConnection) and back it up with a comment why it could fail?